cpp/src/reshape/table_to_array.cu

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/reshape.hpp>
+#include <cudf/reshape.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <cub/device/device_memcpy.cuh>
+#include <cuda/functional>
+#include <cuda_runtime.h>
+#include <thrust/device_vector.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+namespace cudf {
+namespace detail {
+namespace {
+
+template <typename T>
+void _table_to_array(table_view const& input, void* output, rmm::cuda_stream_view stream)
+{
+  auto const num_columns = input.num_columns();
+  auto const num_rows    = input.num_rows();
+  auto const item_size   = sizeof(T);
+  auto* base_ptr         = static_cast<cuda::std::byte*>(output);
+
+  CUDF_EXPECTS(num_columns > 0, "Must have at least one column.");
+  CUDF_EXPECTS(output != nullptr, "Output pointer cannot be null.");
+
+  rmm::device_uvector<void*> d_srcs(num_columns, stream);
+  rmm::device_uvector<void*> d_dsts(num_columns, stream);
+
+  std::vector<void const*> h_srcs(num_columns);
+  std::vector<void*> h_dsts(num_columns);
+
+  for (int i = 0; i < num_columns; ++i) {
+    auto const& col = input.column(i);
+    CUDF_EXPECTS(col.type() == input.column(0).type(), "All columns must have the same dtype");
+    CUDF_EXPECTS(col.null_count() == 0, "All columns must be non-nullable or contain no nulls");
+
+    h_srcs[i] = static_cast<void const*>(col.data<T>());
+    h_dsts[i] = static_cast<void*>(base_ptr + i * item_size * num_rows);
+  }
+
+  CUDF_CUDA_TRY(cudaMemcpyAsync(d_srcs.data(),
+                                h_srcs.data(),
+                                sizeof(void*) * num_columns,
+                                cudaMemcpyHostToDevice,
+                                stream.value()));
+  CUDF_CUDA_TRY(cudaMemcpyAsync(d_dsts.data(),
```

Comment:
I could use some help on how to avoid these raw memcpy calls.

Reply:
Thanks. Done in https://github.com/rapidsai/cudf/pull/18450/commits/0343bb3197e8579aa610898f49cda82423c863d6